core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1155,6 +1156,10 @@ public PlanWithProperties visitUnion(UnionNode node, PreferredProperties parentP
 
             for (int i = 0; i < node.getSources().size(); i++) {
                 PlanWithProperties child = node.getSources().get(i).accept(this, PreferredProperties.any());
+                if (QueryCardinalityUtil.isEmpty(child.getNode(), Lookup.noLookup())) {
```

Comment:
This should be handled by the `RemoveEmptyUnionBranches` rule. It looks like we're missing an instance of this rule in one of the phases before AddExchanges. We should identify where the empty Values is being produced and add the rule there.

Reply:
Hi, @martint thank you for pointing this out, Yes I agree with you! **but** seems like when the rule fires the CBO has not yet resolved the filter pushdown into table scan to prune out the empty partitions. 
**A couple of things:**
1. my read on this [thread](https://trinodb.slack.com/archives/CP1MUNEUX/p1627402534296400?thread_ts=1627314386.283600&cid=CP1MUNEUX) is that adding another pass/call to iterative optimizer is expensive. 
2. In your opinion why do you think this patch is a bad idea (at least from the curiosity perspective)? 
3. Can you please tell me why the CBO is adding a single gather node when we have a mix of Single scans and partitioned scans why can't we have a shuffle to keep parallelism at least when parents prefer distributed nodes
4. Finally seems like we have a major costing issue as you can see in the [issue/examples](https://github.com/trinodb/trino/issues/9981) I have added, the CBO need to be able to avoid a single node Joins and revert back to shuffle join in this case IMO we need to have an overall plan cost and use it to pick plans with better cost.